### PR TITLE
Support for Redis Chat History Buffer

### PR DESCRIPTION
**Pull request by denen99**
# Description

This PR includes the ability to use Redis for the chat history to ensure persistence. Particularly when running RAG applications in environments like AWS Lambda, memory based history is insufficient. Additionally if you want to avoid memory issues with a lot of users and large histories, and be able to scale out your application to multiple nodes decoupling the chat history from the node, then you need an external cache/storage system to handle this

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/memory/redis_chat_memory_buffer.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
+import json
+import logging
+import sys
+from typing import Any, Callable, List, Optional
+from urllib.parse import urlparse
+
+from redis import Redis, RedisCluster
```
Comment (logan-markewich) [inline llama_index/memory/redis_chat_memory_buffer.py:7]: This should be in a try/except import block, inside the functions where it is used

If you are using redis for type checking, you can keep this import too, but put it under a check

```
from typing import TYPE_CHECKING, ...

if TYPE_CHECKING:
    from redis import Redis, RedisCluster
```

**Review by logan-markewich — COMMENTED** [review]
Nice, a clean PR! Thanks for fixing that up.

I think I still want to refactor this into a base memory storage class though. Let me take a stab at refactoring on this branch tomorrow

**Comment by logan-markewich** [comment]
Heads up, finally created a base class for this over here :) https://github.com/run-llama/llama_index/pull/9863

So the idea would be that redis can extend `BaseChatStore`

**Comment by denen99** [comment]
awesome!  would you prefer i move my commits over to a fork of that PR ? Or we get that PR merged first and then i can merge those changes into this PR ?

**Comment by logan-markewich** [comment]
@denen99 the PR is merged. I think you can either merge main into this PR branch and update, or you can open a new PR

The work will largely be similar, but you'll want to implement the BaseChatStore class now 👍

**Comment by denen99** [comment]
OK.  I am going to close this PR and open a new one purely based on the RedisChatStore given the new implementation to avoid confusion, thank you!

**Pull request closed without merging**